llvm/lib/Target/RISCV/RISCVTargetObjectFile.cpp

```diff
@@ -104,10 +104,34 @@ bool RISCVELFTargetObjectFile::isGlobalInSmallSection(
 MCSection *RISCVELFTargetObjectFile::SelectSectionForGlobal(
     const GlobalObject *GO, SectionKind Kind, const TargetMachine &TM) const {
   // Handle Small Section classification here.
-  if (Kind.isBSS() && isGlobalInSmallSection(GO, TM))
-    return SmallBSSSection;
-  if (Kind.isData() && isGlobalInSmallSection(GO, TM))
-    return SmallDataSection;
+  if (isGlobalInSmallSection(GO, TM)) {
+    // Emit to an unique sdata/sbss section when -fdata-section is set.
+    // However, if a symbol has an explicit sdata/sbss section, place it in that
+    // section.
+    bool EmitUniquedSection = TM.getDataSections() && !GO->hasSection();
```

Comment:
I think this is still wrong, and that the `!GO->hasSection()` check needs to take priority over the `isGlobalInSmallSection` - I noted in the test where I think things will still go wrong.

Reply:
I think `isGlobalInSmallSection` always return `false` unless the explicit section is a `.sdata` or `.sbss`, so I meant to let symbols with `__attribute__((section(".sdata")))`/`__attribute__((section(".sbss")))` neglect -fdata-sections and directly use `SmallDataSection`/`SmallBSSSection`.
Please correct me if I misunderstand, thanks.

Reply:
Ah, I see what's going on.